examples/vision/forwardforward.py

```diff
@@ -0,0 +1,352 @@
+
+                loss = tf.math.log(
+                    1
+                    + tf.math.exp(
+                        tf.concat([-g_pos + self.threshold, g_neg - self.threshold], 0)
+                    )
+                )
+                mean_loss = tf.math.reduce_mean(loss)
+                loss_list.append(mean_loss.numpy())
+            gradients = tape.gradient(mean_loss, self.trainable_weights)
+            self.optimizer.apply_gradients(zip(gradients, self.dense.trainable_weights))
+        return (
+            tf.stop_gradient(self.call(x_pos)),
+            tf.stop_gradient(self.call(x_neg)),
+            mean_loss,
+            loss_list,
+        )
+
+
+"""
+## Define the `FFNetwork` Custom Model
+
+With our custom layer defined, we also need to override the `train_step` method and
+define a custom `tf.keras.models.Model` that works with our `FFDense` layer.
+
+For this algorithm, we must 'embed' the labels onto the original image. To do so, we
+exploit the structure of MNIST images where the top-left 10 pixels are always zeros. We
+use that as a label space in order to visually one-hot-encode the labels within the image
+itself. This action is performed by the `overlay_y_on_x` function.
+
+We break down the prediction function with a per-sample prediction function which is then
+called over the entire test set by the overriden `predict()` function. The prediction is
+performed here with the help of measuring the `excitation` of the neurons per layer for
+each image. This is then summed over all layers to calculate a network-wide 'goodness
+score'. The label with the highest 'goodness score' is then chosen as the sample
+prediction.
+
+The `train_step` function is overriden to act as the main controlling loop for running
+training on each layer as per the number of epochs per layer.
+"""
+
+
+class FFNetwork(keras.Model):
+    def __init__(self, dims, layer_optimizer=keras.optimizers.Adam(learning_rate=0.03)):
+        super().__init__()
+        self.layer_optimizer = layer_optimizer
+        self.mean_loss = keras.metrics.Mean()
+        self.flatten_layer = keras.layers.Flatten()
+        self.layer_list = [keras.Input(shape=(dims[0],))]
+        for d in range(len(dims) - 1):
+            self.layer_list += [FFDense(dims[d + 1], optimizer=self.layer_optimizer)]
+
+    @tf.function()
+    def overlay_y_on_x(self, X, y):
+        x_res = X.numpy()
+        x_npy = X.numpy()
+        x_res[:, :10] *= 0.0
+        if not isinstance(y, int):
+            y_npy = y.numpy()
+            x_res[range(x_npy.shape[0]), y.numpy()] = x_npy.max()
+        else:
+            x_res[range(x_npy.shape[0]), y] = x_npy.max()
+        return tf.convert_to_tensor(x_res)
+
+    @tf.function()
+    def predict_one_sample(self, x):
+        goodness_per_label = []
+        x = tf.expand_dims(x, axis=0)
+        for label in range(10):
+            h = self.overlay_y_on_x(x, label)
+            h = self.flatten_layer(h)
```

Comment:
It's a bit weird to use a layer for this, you can just use a `tf.reshape([tf.shape(h), -1])` op